courses/codepeer/010_codepeer.rst

```diff
@@ -0,0 +1,2666 @@
+
+   procedure Div is
+      type Int is range 0 .. 2**32 - 1;
+      A : Int := Int'Last;
+      X : Integer;
+   begin
+      for I in Int range 0 .. 2 loop
+         X := Integer (A / I); -- division by zero when I=0
+      end loop;
+   end Div;
+
+| ``high: divide by zero fails here: requires I /= 0``
+
+--------------
+Access Check
+--------------
+
++ Attempting to dereference a reference that could be :ada:`null`
++ Will generate an :ada:`Access_Error`
+
+..
+   :toolname:`CodePeer` example (4.1.1 - access check)
+
+.. code:: Ada
+   :number-lines: 1
+
+   procedure Null_Deref is
+      type Int_Access is access Integer;
+      X : Int_Access;
+   begin
+      if X = null then
+         X.all := 1;  -- null dereference
+      end if;
+   end Null_Deref;
+
+| ``high: access check fails here``
+
+-------------
+Range Check
+-------------
+
++ Calculation may generate a value outside the :ada:`range` of an Ada type or subtype
++ Will generate a :ada:`Constraint_Error`
+
+..
+   :toolname:`CodePeer` example (4.1.1 - range check)
+
+.. code:: Ada
+   :number-lines: 1
+
+   subtype Constrained_Integer is Integer range 1 .. 2;
+   A : Integer;
+
+   procedure Proc_1 (I : in Constrained_Integer) is
+   begin
+      A := I + 1;
+   end Proc_1;
+   ...
+   A := 0;
+   Proc_1 (I => A);  --  A is out-of-range of parameter I
+
+| ``high: range check fails here: requires A in 1..2``
+
+----------------
+Overflow Check
+----------------
+
++ Calculation may overflow the bounds of a numeric type.
++ Depends on the size of the underlying type (base type)
++ Will generate a :ada:`Constraint_Error`
+
+..
+   :toolname:`CodePeer` example (4.1.1 - overflow check)
+
+.. code:: Ada
+   :number-lines: 1
+
+   is
+      Attempt_Count : Integer := Integer'Last;
+   begin
+      -- Forgot to reset Attempt_Count to 0
+      loop
+         Put ("Enter password to delete system disk");
+         if Get_Correct_Pw then
+            Allow_Access;
+         else
+            Attempt_Count := Attempt_Count + 1;
+
+| ``high: overflow check fails here: requires Attempt_Count /= Integer_32'Last``
+| ``high: overflow check fails here: requires Attempt_Count in Integer_32'First-1..Integer_32'Last-1``
+
+----------------
+Aliasing Check
+----------------
+
++ Some parameters could be passed as **reference**
++ For such parameters, **precondition** that they:
+
+  + Are a **different object** from other parameters
+  + Do **not match** the address of a global object
+
+..
+   :toolname:`CodePeer` example (4.1.1 - aliasing check)
+
+.. code:: Ada
+   :number-lines: 1
+
+      procedure In_Out (A : Int_Array; B : out Int_Array) is
+      begin
+         B (1) := A (1) + 1;
+         ...
+         B (1) := A (1) + 2;
+      end In_Out;
+   ...
+      In_Out (A, A); -- Aliasing!
+
+| ``high: precondition (aliasing check) failure on call to alias.in_out: requires B /= A``
+
+-----------
+Tag Check
+-----------
+
+A tag check operation on a :ada:`tagged` object might fail
```

Comment:
"Will generate Constraint_Error"

Reply:
"Tagged" isn't an Ada keyword, so probably should use :dfn:

Reply:
> "Tagged" isn't an Ada keyword, so probably should use :dfn:

? it is though